File: 2021/golang/src/day2/main.go
package main

import (
	"strconv"
	"strings"
	"utils"
)

var data = getData(utils.GetLines("input.txt"))

func getData(lines []string) [][]string {
	var retVal [][]string
	for _, e := range lines {
		retVal = append(retVal, strings.Split(e, " "))
	}
	return retVal
}

func partOne() string {
	dist := 0
	depth := 0
	for _, d := range data {
		i, _ := strconv.Atoi(d[1])
		if d[0] == "forward" {
			dist += i
		} else if d[0] == "down" {
			depth += i
		} else if d[0] == "up" {
			depth -= i
		}
	}

	return strconv.Itoa(depth * dist)
}

func partTwo() string {
	dist := 0
	depth := 0
	aim := 0
	for _, d := range data {
		i, _ := strconv.Atoi(d[1])
		if d[0] == "forward" {
			dist += i
			depth += aim * i
		} else if d[0] == "down" {
			aim += i
		} else if d[0] == "up" {
			aim -= i
		}
	}

	return strconv.Itoa(depth * dist)
}

func main() {
	utils.RunWithTimer(partOne) // 1660158 -- took 0 ms
	utils.RunWithTimer(partTwo) // 1604592846 -- took 0 ms
}

File: 2021/python/day2/main.py
from aoc_utils import run_with_timer

data = [x.strip().split() for x in open("input.txt").readlines()]


def part_one():
	dist = sum(int(d[1]) for d in data if d[0] == 'forward')
	depth = sum(int(d[1]) for d in data if d[0] == 'down') - sum(int(d[1]) for d in data if d[0] == 'up')
	return dist * depth


def part_two():
	curr_depth = 0
	curr_dist = 0
	curr_aim = 0
	for d in data:
		match d[0]:
			case 'forward':
				curr_dist += int(d[1])
				curr_depth += curr_aim * int(d[1])
			case 'down':
				curr_aim += int(d[1])
			case 'up':
				curr_aim -= int(d[1])
	return curr_depth * curr_dist


if __name__ == '__main__':
	run_with_timer(part_one)  # 1660158 -- took 0 ms
	run_with_timer(part_two)  # 1604592846 -- took 0 ms
